kotlin/workflow-runtime/src/main/java/com/squareup/workflow/internal/WorkflowLoop.kt

```diff
@@ -27,79 +27,94 @@ import kotlinx.coroutines.flow.Flow
+  suspend fun <InputT, StateT, OutputT : Any, RenderingT> runWorkflowLoop(
+    workflow: StatefulWorkflow<InputT, StateT, OutputT, RenderingT>,
+    inputs: Flow<InputT>,
+    initialSnapshot: Snapshot?,
+    initialState: StateT? = null,
+    onRendering: suspend (RenderingAndSnapshot<RenderingT>) -> Unit,
+    onOutput: suspend (OutputT) -> Unit
+  ): Nothing
+}
+
+internal object RealWorkflowLoop : WorkflowLoop {
+
+  @UseExperimental(FlowPreview::class, ExperimentalCoroutinesApi::class)
+  override suspend fun <InputT, StateT, OutputT : Any, RenderingT> runWorkflowLoop(
+    workflow: StatefulWorkflow<InputT, StateT, OutputT, RenderingT>,
+    inputs: Flow<InputT>,
+    initialSnapshot: Snapshot?,
+    initialState: StateT?,
+    onRendering: suspend (RenderingAndSnapshot<RenderingT>) -> Unit,
+    onOutput: suspend (OutputT) -> Unit
+  ): Nothing = coroutineScope {
+    val inputsChannel = inputs.produceIn(this)
+    inputsChannel.consume {
+      var output: OutputT? = null
+      var input: InputT = inputsChannel.receive()
+      var inputsClosed = false
+      val rootNode = WorkflowNode(
+          id = workflow.id(),
+          workflow = workflow,
+          initialInput = input,
+          snapshot = initialSnapshot,
+          baseContext = coroutineContext,
+          initialState = initialState
+      )
 
-    try {
-      while (true) {
-        coroutineContext.ensureActive()
+      try {
+        while (true) {
+          coroutineContext.ensureActive()
 
-        val rendering = rootNode.render(workflow, input)
-        val snapshot = rootNode.snapshot(workflow)
+          val rendering = rootNode.render(workflow, input)
+          val snapshot = rootNode.snapshot(workflow)
 
-        onRendering(RenderingAndSnapshot(rendering, snapshot))
-        output?.let { onOutput(it) }
+          onRendering(RenderingAndSnapshot(rendering, snapshot))
+          output?.let { onOutput(it) }
 
-        // Tick _might_ return an output, but if it returns null, it means the state or a child
-        // probably changed, so we should re-render/snapshot and emit again.
-        output = select {
-          // Stop trying to read from the inputs channel after it's closed.
-          if (!inputsClosed) {
-            @Suppress("EXPERIMENTAL_API_USAGE")
-            inputsChannel.onReceiveOrNull { newInput ->
-              if (newInput == null) {
-                inputsClosed = true
-              } else {
-                input = newInput
+          // Tick _might_ return an output, but if it returns null, it means the state or a child
+          // probably changed, so we should re-render/snapshot and emit again.
+          output = select {
+            // Stop trying to read from the inputs channel after it's closed.
+            if (!inputsClosed) {
+              @Suppress("EXPERIMENTAL_API_USAGE")
+              inputsChannel.onReceiveOrNull { newInput ->
+                if (newInput == null) {
+                  inputsClosed = true
```

Comment:
Have we documented that closing inputs does not end the workflow session?

Reply:
We did at one point. I'll add it to the docs on `launchWorkflowIn`.

- [ ] Document input-closing behavior in `launchWorkflowIn`